erigon-lib/p2p/sentry/sentrymultiplexer.go

```diff
@@ -0,0 +1,614 @@
+			g.Go(func() error {
+				_, err := client.SetStatus(gctx, in, opts...)
+				return err
+			})
+		}
+	}
+
+	err := g.Wait()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &sentryproto.SetStatusReply{}, nil
+}
+
+func (m *sentryMultiplexer) PenalizePeer(ctx context.Context, in *sentryproto.PenalizePeerRequest, opts ...grpc.CallOption) (*emptypb.Empty, error) {
+	g, gctx := errgroup.WithContext(ctx)
+
+	for _, client := range m.clients {
+		client := client
+
+		g.Go(func() error {
+			_, err := client.PenalizePeer(gctx, in, opts...)
+			return err
+		})
+	}
+
+	return &emptypb.Empty{}, g.Wait()
+}
+
+func (m *sentryMultiplexer) PeerMinBlock(ctx context.Context, in *sentryproto.PeerMinBlockRequest, opts ...grpc.CallOption) (*emptypb.Empty, error) {
+	g, gctx := errgroup.WithContext(ctx)
+
+	for _, client := range m.clients {
+		client := client
+
+		g.Go(func() error {
+			_, err := client.PeerMinBlock(gctx, in, opts...)
+			return err
+		})
+	}
+
+	return &emptypb.Empty{}, g.Wait()
+}
+
+// Handshake is not performed on the multi-client level
+func (m *sentryMultiplexer) HandShake(ctx context.Context, in *emptypb.Empty, opts ...grpc.CallOption) (*sentryproto.HandShakeReply, error) {
+	g, gctx := errgroup.WithContext(ctx)
+
+	var protocol sentryproto.Protocol
+	var mu sync.Mutex
+
+	for _, client := range m.clients {
+		client := client
+
+		if !client.hasHandshake {
+			g.Go(func() error {
+				reply, err := client.HandShake(gctx, &emptypb.Empty{}, grpc.WaitForReady(true))
+
+				if err != nil {
+					return err
+				}
+
+				mu.Lock()
+				defer mu.Unlock()
+
+				if reply.Protocol > protocol {
+					protocol = reply.Protocol
+				}
+
+				client.hasHandshake = true
+
+				return nil
+			})
+		}
+	}
+
+	err := g.Wait()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &sentryproto.HandShakeReply{Protocol: protocol}, nil
+}
+
+func (m *sentryMultiplexer) SendMessageByMinBlock(ctx context.Context, in *sentryproto.SendMessageByMinBlockRequest, opts ...grpc.CallOption) (*sentryproto.SentPeers, error) {
+	g, gctx := errgroup.WithContext(ctx)
+
+	var allSentPeers []*typesproto.H512
+	var allSentMutex sync.RWMutex
+
+	for _, client := range m.clients {
+		client := client
+
+		g.Go(func() error {
+			sentPeers, err := client.SendMessageByMinBlock(gctx, in, opts...)
+
+			if err != nil {
+				return err
+			}
+
+			allSentMutex.Lock()
+			defer allSentMutex.Unlock()
+
+			allSentPeers = append(allSentPeers, sentPeers.GetPeers()...)
+
+			return nil
+		})
+	}
+
+	err := g.Wait()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &sentryproto.SentPeers{Peers: allSentPeers}, nil
+}
+
+func (m *sentryMultiplexer) SendMessageById(ctx context.Context, in *sentryproto.SendMessageByIdRequest, opts ...grpc.CallOption) (*sentryproto.SentPeers, error) {
+	g, gctx := errgroup.WithContext(ctx)
+
+	var allSentPeers []*typesproto.H512
+	var allSentMutex sync.RWMutex
+
+	for _, client := range m.clients {
+		client := client
+
+		g.Go(func() error {
```

Comment:
I'm not sure if the sends should send to all sentries in parallel because we send a message to a peer using its pub key as id and the same peer may have more than 1 sentry running for each protocol.

For example if another Erigon peer is running 2 sentries for eth68 and eth67 that means we will hit that peer with 2 requests instead of 1 and it will have to respond twice for the same reply. 

This may be a problem for requests like `GetBlockBodies` (we send these for ranges of 1024 blocks for checkpoints) - the peer will have to respond twice instead of once and the responses to these requests are not light.

I'm also noticing that in the current SentryMultiClient implementation it sends to the sentries in a sequential order -> checks if the send succeeded -> if it did it stops, while if it did not succeed it moves on to the next sentry - https://github.com/erigontech/erigon/blob/main/p2p/sentry/sentry_multi_client/sentry_api.go#L88-L91